src/FakeItEasy/Configuration/AfterMustHaveHappenedConfiguration.cs

```diff
@@ -17,11 +17,38 @@ public AfterMustHaveHappenedConfiguration(FakeManager fakeManager, ICallMatcher
             this.repeatConstraint = repeatConstraint;
         }
 
-        public void InOrder(ISequentialCallContext context)
+        /// <summary>
+        /// Checks that the asserted call happened in order relative to others in the assertion chain.
+        /// </summary>
+        /// <param name="nextAssertion">An assertion describing the next call that should occur.</param>
+        /// <returns>An object that can be used to assert that a following call was made in the expected order.</returns>
+        /// <exception cref="ExpectationException">The call was not made in the expected order.</exception>
+        public IOrderableCallAssertion Then(IAfterMustHaveHappenedConfiguration nextAssertion)
         {
-            Guard.AgainstNull(context, "context");
+            var context = ServiceLocator.Current.Resolve<SequentialCallContext>();
+            this.CheckCallHappenedInOrder(context);
+            return new SubsequentOrderedCallAssertion(context).Then(nextAssertion);
+        }
 
+        private void CheckCallHappenedInOrder(SequentialCallContext context)
+        {
             context.CheckNextCall(this.fakeManager, this.matcher.Matches, this.callDescription, this.repeatConstraint);
         }
+
+        private class SubsequentOrderedCallAssertion : IOrderableCallAssertion
+        {
+            private readonly SequentialCallContext context;
+
+            public SubsequentOrderedCallAssertion(SequentialCallContext context)
+            {
+                this.context = context;
+            }
+
+            public IOrderableCallAssertion Then(IAfterMustHaveHappenedConfiguration nextCall)
+            {
+                ((AfterMustHaveHappenedConfiguration)nextCall).CheckCallHappenedInOrder(this.context);

```

Comment:
> As noted before, we could return the concrete class. I'm not sure what's so terrible about this.

It's not terrible; it's just inconsistent with what we do elsewhere in the API, which bothers me slightly, but isn't a very serious problem.

An abstract class with a protected constructor would be fine. I would suggest a different name, though; `AfterMustHaveHappenedConfiguration` doesn't fit very well IMO, because it's not used to configure anything, but only to chain assertions. The way I see it, what this class really represents is a successful assertion (because if the assertion fails regardless of the order, `MustHaveHappened()` will throw). So we could call it `AssertedCall`, or `SuccessfullCallAssertion`, or something along those lines.

> 1. comments explaining why we're casting and not to add additional production-ready implementations of IAfterMustHaveHappenedConfiguration
> 2. to augment the comments, we could even add a unit test to ensure that nobody adds an implementation unawares

This doesn't really help; it could prevent _us_ from accidentally adding other implementations in FakeItEasy itself, but it wouldn't prevent the user from doing so. Since the interface is public, anyone can implement it and pass an instance of their implementation to `Then`. I hadn't realized that initially, but now that I see it, I agree that the cast is an issue; it's a violation of the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle).